From c0c32d8403638358e1a97db16dee9cd3fe6d9f7a Mon Sep 17 00:00:00 2001
From: Adwait Nayak <quic_adwaitn@quicinc.com>
Date: Tue, 8 Oct 2024 17:56:20 +0530
Subject: [PATCH] UPSTREAM: CP: wifi: ath12k: Add module param control for CFR

Usually CFR feature requires below memory allocations:
DBR allocation :
Number of buffers * Max buf size
In case of WKK, it is 103 * 64KB = 6.5 MB per radio
cfr_test_app : Allocating 8MB file per radio in /tmp/
relayfs: Allocates approax. 1MB file per radio

Out of which, cfr_test_app and relayfs memories are allocated
during runtime, whereas the DBR allocation is done even if
CFR is disabled; which is an overhead.

Fix:
- Introduction of new module param named cfr_enable_bmap.
- By default cfr_enable_bmap would be set to 0, indicating that
  CFR feature is not enabled by default. It's a bitmap.
  User can set the bitmap in /etc/modules.d/ath12k in order
  to enable/ disable CFR per radio.
  Ex: If ath12k cfr_enable_bmap = 0x7, CFR would be enabled for
  chip_id 0/1/2. If ath12k cfr_enable_bmap = 0x3, CFR would be
  enabled for chip_id 0/1 etc.

- Skip CFR SRNG allocations
- Skip WMI handler registrations for DBR/TX completion events
- Skip relay write (cfr_capture will not be created)
- If CFR is disabled, it will not allow user to start
  cfr_test_app as well.

Patch-dependency: 890-ath12k-add-CFR-single-shot-periodic-capture-support.patch
Patch-work: None
Fixes: None

Signed-off-by: Adwait Nayak <quic_adwaitn@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -118,6 +118,10 @@ module_param_named(rx_nwifi_err_dump, at
 MODULE_PARM_DESC(rx_nwifi_err_dump,
 		 "rx nwifi err dump: 0-disable, 1-enable");
 
+unsigned int ath12k_cfr_enable_bmap = 0;
+module_param_named(cfr_enable_bmap, ath12k_cfr_enable_bmap, uint, 0644);
+MODULE_PARM_DESC(cfr_enable_bmap, "cfr_enable_bmap: 0-disable, enable-(0x7 for 3 chipsets)");
+
 static DEFINE_MUTEX(ath12k_hw_lock);
 static struct list_head ath12k_hw_groups = LIST_HEAD_INIT(ath12k_hw_groups);
 
@@ -1401,10 +1405,14 @@ static int ath12k_core_pdev_init(struct
 		goto err_spectral_deinit;
 	}
 
-	ret = ath12k_cfr_init(ab);
-	if (ret) {
-		ath12k_err(ab, "failed to init cfr %d\n", ret);
-		goto err_spectral_deinit;
+	/* Check if cfr_enable_bmap is set for the corresponding HW */
+	if (ath12k_cfr_enable_bmap & (1 << ab->chip_id)) {
+		ath12k_info(ab, "Enabling CFR for chip id:%d\n", ab->chip_id);
+		ret = ath12k_cfr_init(ab);
+		if (ret) {
+			ath12k_err(ab, "failed to init cfr %d\n", ret);
+			goto err_spectral_deinit;
+		}
 	}
 
 	ath12k_sawf_init(ab);
@@ -1432,7 +1440,9 @@ err_thermal_unregister:
 static void ath12k_core_pdev_deinit(struct ath12k_base *ab)
 {
 	ath12k_dp_ppeds_stop(ab);
-	ath12k_cfr_deinit(ab);
+	if (ath12k_cfr_enable_bmap & (1 << ab->chip_id))
+		ath12k_cfr_deinit(ab);
+
 	ath12k_spectral_deinit(ab);
 	ath12k_thermal_unregister(ab);
 	ath12k_dp_accel_cfg_deinit(ab);
--- a/drivers/net/wireless/ath/ath12k/hw.c
+++ b/drivers/net/wireless/ath/ath12k/hw.c
@@ -211,6 +211,8 @@ void ath12k_hw_ipq5332_fill_cfr_hdr_info
 
 	if(ar->ab->hw_rev == ATH12K_HW_QCN6432_HW10)
 		header->chip_type = ATH12K_CFR_RADIO_QCN6432;
+	else if (ar->ab->hw_rev == ATH12K_HW_IPQ5424_HW10)
+		header->chip_type = ATH12K_CFR_RADIO_IPQ5424;
 	else
 		header->chip_type = ATH12K_CFR_RADIO_IPQ5332;
 
@@ -2405,16 +2407,20 @@ static struct ath12k_hw_params ath12k_hw
 		.is_plink_preferable = true,
 		.support_fse = true,
 		.handle_beacon_miss = true,
-		.cfr_support = false,
+		.cfr_support = true,
 		.ds_support = true,
 		.cfr_dma_hdr_size = sizeof(struct ath12k_cfir_enh_dma_hdr),
 		.cfr_num_stream_bufs = 128,
-		/* sizeof (ath12k_csi_cfr_header) + max cfr header(200 bytes) +
-		 * max cfr payload(16384 bytes)
+		/* Max size :
+		 * sizeof(ath12k_csi_cfr_header) + 96 bytes(cfr header) +
+		 *				15744 bytes(cfr payload)
+		 * where cfr_header = rtt upload header len +
+		 *                    freeze_tlv len + uplink user setup info
+		 *                  = 32bytes + 32bytes + (8bytes * 4users)
 		 */
 		.cfr_stream_buf_size = sizeof(struct ath12k_csi_cfr_header) +
-					(CFR_HDR_MAX_LEN_WORDS_QCN6432 * 4) +
-					CFR_DATA_MAX_LEN_QCN6432,
+					(CFR_HDR_MAX_LEN_WORDS_IPQ5424 * 4) +
+					CFR_DATA_MAX_LEN_IPQ5424,
 	},
 };
 
--- a/drivers/net/wireless/ath/ath12k/cfr.h
+++ b/drivers/net/wireless/ath/ath12k/cfr.h
@@ -25,19 +25,23 @@
 #define ATH12K_CFR_RADIO_WCN7850 33
 #define ATH12K_CFR_RADIO_IPQ5332 35
 #define ATH12K_CFR_RADIO_QCN6432 38
+#define ATH12K_CFR_RADIO_IPQ5424 42
 
-#define CFR_HDR_MAX_LEN_WORDS_QCN9274 88
+#define CFR_HDR_MAX_LEN_WORDS_QCN9274 90
 #define CFR_DATA_MAX_LEN_QCN9274 64512
 
 #define CFR_HDR_MAX_LEN_WORDS_WCN7850 16
 #define CFR_DATA_MAX_LEN_WCN7850 16064
 
-#define CFR_HDR_MAX_LEN_WORDS_IPQ5332 22
+#define CFR_HDR_MAX_LEN_WORDS_IPQ5332 24
 #define CFR_DATA_MAX_LEN_IPQ5332 8192
 
-#define CFR_HDR_MAX_LEN_WORDS_QCN6432 30
+#define CFR_HDR_MAX_LEN_WORDS_QCN6432 32
 #define CFR_DATA_MAX_LEN_QCN6432 32152
 
+#define CFR_HDR_MAX_LEN_WORDS_IPQ5424 24
+#define CFR_DATA_MAX_LEN_IPQ5424 15744
+
 #define VENDOR_QCA 0x8cfdf0
 #define PLATFORM_TYPE_ARM 2
 #define NUM_CHAINS_FW_TO_HOST(n) ((1 << ((n) + 1)) - 1)
@@ -105,7 +109,7 @@ struct ath12k_cfr_peer_tx_param {
 };
 
 #define HOST_MAX_CHAINS 8
-#define MAX_CFR_MU_USERS 4
+#define MAX_CFR_MU_USERS 37
 
 struct cfr_dbr_metadata {
 	u8 peer_addr[ETH_ALEN];
@@ -230,16 +234,199 @@ struct ath12k_cfir_dma_hdr {
 #define CFIR_DMA_HDR_INFO3_DECIMATION_FACT GENMASK(11, 8)
 #define CFIR_DMA_HDR_INFO3_RSVD GENMASK(15, 12)
 
+/*
+ * @tag: ucode fills this with 0xBA
+ *
+ * @length: length of CFR header in words (32-bit)
+ *
+ * @upload_done: ucode sets this to 1 to indicate DMA completion
+ *
+ * @capture_type:
+ *
+ *                      0 - None
+ *                      1 - RTT-H (Nss = 1, Nrx)
+ *                      2 - Debug-H (Nss, Nrx)
+ *                      3 - Reserved
+ *                      5 - RTT-H + CIR(Nss, Nrx)
+ *
+ * @preamble_type:
+ *
+ *                      0 - Legacy
+ *                      1 - HT
+ *                      2 - VHT
+ *                      3 - HE
+ *
+ * @nss:
+ *
+ *                      0 - 1-stream
+ *                      1 - 2-stream
+ *                      ..      ..
+ *                      7 - 8-stream
+ *
+ *@num_chains:
+ *
+ *                      0 - 1-chain
+ *                      1 - 2-chain
+ *                      ..  ..
+ *                      7 - 8-chain
+ *
+ *@upload_bw_pkt:
+ *
+ *                      0 - 20 MHz
+ *                      1 - 40 MHz
+ *                      2 - 80 MHz
+ *                      3 - 160 MHz
+ *
+ * @sw_peer_id_valid: Indicates whether sw_peer_id field is valid or not,
+ * sent from MAC to PHY via the MACRX_FREEZE_CAPTURE_CHANNEL TLV
+ *
+ * @sw_peer_id: Indicates peer id based on AST search, sent from MAC to PHY
+ * via the MACRX_FREEZE_CAPTURE_CHANNEL TLV
+ *
+ * @phy_ppdu_id: sent from PHY to MAC, copied to MACRX_FREEZE_CAPTURE_CHANNEL
+ * TLV
+ *
+ * @total_bytes: Total size of CFR payload (FFT bins)
+ *
+ * @header_version:
+ *
+ *                      1 - IPQ87XX
+ *                      2 - IPQ6018
+ *                      3 - 11BE chipsets
+ *
+ * @target_id:
+ * @cfr_fmt:
+ *
+ *                      0 - raw (32-bit format)
+ *                      1 - compressed (24-bit format)
+ *
+ * @mu_rx_data_incl: Indicates whether CFR header contains UL-MU-MIMO info
+ *
+ * @freeze_data_incl: Indicates whether CFR header contains
+ * MACRX_FREEZE_CAPTURE_CHANNEL TLV
+ *
+ * @freeze_tlv_version: Indicates the version of freeze_tlv
+ *                      1 - IPQ87xx, IPQ6018
+ *                      2 - IPQ5018
+ *                      3 - IPQ9000
+ *                      5 - 11BE chipsets
+ *
+ * @decimation_factor: FFT bins decimation
+ * @mu_rx_num_users: Number of users in UL-MU-PPDU
+ * @he_ltf_type:
+ * @ext_preamble_type:
+ * @amplitude_gain_ratio_0_3:
+ * @rescale_amt_shift_pri80:
+ * @rescale_amt_shift_sec80:
+ * @cgim_status:
+ * @cgim_filter:
+ * @phy_mode:
+ * @demf_turbo_mode:
+ * @demf_pbs_en:
+ * @leg_cfr_mode:
+ * @puncture_pattern:
+ * @pri20_location:
+ * @channel_bandwidth:
+ * @_11az_mode:
+ * @_11az_node:
+ */
 struct ath12k_cfir_enh_dma_hdr {
-	struct ath12k_cfir_dma_hdr hdr;
+	u16 tag              :  8,
+	    length           :  6,
+	    rsvd1            :  2;
+	u16 upload_done        :  1,
+	    capture_type       :  3,
+	    preamble_type      :  2,
+	    nss                :  3,
+	    num_chains         :  3,
+	    upload_pkt_bw      :  3,
+	    sw_peer_id_valid   :  1;
+	u16 sw_peer_id         : 16;
+	u16 phy_ppdu_id        : 16;
 	u16 total_bytes;
-	u16 info2;
-	u16 info3;
-	u16 rsvd;
+	u16 header_version     :4,
+	    target_id          :4,
+	    cfr_fmt            :1,
+	    cir_fmt            :1,
+	    mu_rx_data_incl    :1,
+	    freeze_data_incl   :1,
+	    freeze_tlv_version :4;
+	u16 mu_rx_num_users    :8,
+	    decimation_factor  :4,
+	    he_ltf_type        :4;
+	u16 ext_preamble_type  :1,
+	    rsvd2              :15;
+	u32 amplitude_gain_ratio_0_3;
+	u16 rescale_amt_shift_pri80    : 8,
+	    rescale_amt_shift_sec80    : 8;
+	u16 cgim_status        : 1,
+	    cgim_filter        : 1,
+	    phy_mode           : 1,
+	    demf_turbo_mode    : 1,
+	    demf_pbs_en        : 2,
+	    leg_cfr_mode       : 2,
+	    puncture_pattern   : 8;
+	u16 pri20_location     : 8,
+	    channel_bandwidth  : 3,
+	    _11az_mode         : 4,
+	    _11az_node         : 1;
+	u16 rsvd3;
+	u16 rsvd4;
+	u16 rsvd5;
 };
 
 #define CFR_MAX_LUT_ENTRIES 136
 
+struct macrx_freeze_capture_channel_v5 {
+        u16 freeze                          :  1, //[0]
+            capture_reason                  :  3, //[3:1]
+            packet_type                     :  2, //[5:4]
+            packet_sub_type                 :  4, //[9:6]
+            reserved                        :  5, //[14:10]
+            sw_peer_id_valid                :  1; //[15]
+        u16 sw_peer_id; //[15:0]
+        u16 phy_ppdu_id; //[15:0]
+        u16 packet_ta_lower_16; //[15:0]
+        u16 packet_ta_mid_16; //[15:0]
+        u16 packet_ta_upper_16; //[15:0]
+        u16 packet_ra_lower_16; //[15:0]
+        u16 packet_ra_mid_16; //[15:0]
+        u16 packet_ra_upper_16; //[15:0]
+        u16 tsf_timestamp_15_0; //[15:0]
+        u16 tsf_timestamp_31_16; //[15:0]
+        u16 tsf_timestamp_47_32; //[15:0]
+        u16 tsf_timestamp_63_48; //[15:0]
+        u16 user_index_or_user_mask_5_0     :  6, //[5:0]
+            directed                        :  1, //[6]
+            reserved_13                     :  9; //[15:7]
+        u16 user_mask_21_6; //[15:0]
+        u16 user_mask_36_22                 : 15, //[14:0]
+            reserved_15a                    :  1; //[15]
+};
+
+struct uplink_user_setup_info_v2 {
+        u32 bw_info_valid                   :  1, //[0]
+            uplink_receive_type             :  2, //[2:1]
+            reserved_0a                     :  1, //[3]
+            uplink_11ax_mcs                 :  4, //[7:4]
+            nss                             :  3, //[10:8]
+            stream_offset                   :  3, //[13:11]
+            sta_dcm                         :  1, //[14]
+            sta_coding                      :  1, //[15]
+            ru_type_80_0                    :  4, //[19:16]
+            ru_type_80_1                    :  4, //[23:20]
+            ru_type_80_2                    :  4, //[27:24]
+            ru_type_80_3                    :  4; //[31:28]
+        u32 ru_start_index_80_0             :  6, //[5:0]
+            reserved_1a                     :  2, //[7:6]
+            ru_start_index_80_1             :  6, //[13:8]
+            reserved_1b                     :  2, //[15:14]
+            ru_start_index_80_2             :  6, //[21:16]
+            reserved_1c                     :  2, //[23:22]
+            ru_start_index_80_3             :  6, //[29:24]
+            reserved_1d                     :  2; //[31-30]
+};
+
 struct ath12k_cfr_look_up_table {
 	bool dbr_recv;
 	bool tx_recv;
@@ -287,6 +474,7 @@ enum macrx_freeze_tlv_version {
 	MACRX_FREEZE_TLV_VERSION_1 = 1,
 	MACRX_FREEZE_TLV_VERSION_2 = 2,
 	MACRX_FREEZE_TLV_VERSION_3 = 3,
+	MACRX_FREEZE_TLV_VERSION_5 = 5,
 	MACRX_FREEZE_TLV_VERSION_MAX
 };
 
--- a/drivers/net/wireless/ath/ath12k/cfr.c
+++ b/drivers/net/wireless/ath/ath12k/cfr.c
@@ -245,7 +245,6 @@ static int ath12k_cfr_enh_process_data(s
 	u32 length;
 	u32 freeze_tlv_len = 0;
 	u32 end_magic = ATH12K_CFR_END_MAGIC;
-	u8 freeze_tlv_ver;
 	u8 capture_type;
 	int ret = 0;
 	int status;
@@ -255,18 +254,20 @@ static int ath12k_cfr_enh_process_data(s
 
 	memcpy(&dma_hdr, data, sizeof(struct ath12k_cfir_enh_dma_hdr));
 
-	freeze_tlv_ver = FIELD_GET(CFIR_DMA_HDR_INFO2_FREEZ_TLV_VER, dma_hdr.info2);
-
-	if (FIELD_GET(CFIR_DMA_HDR_INFO2_FREEZ_DATA_INC, dma_hdr.info2)) {
+	if (dma_hdr.freeze_data_incl) {
 		freeze_tlv = data + sizeof(struct ath12k_cfir_enh_dma_hdr);
 		capture_type = freeze_reason_to_capture_type(ab, freeze_tlv);
 	} else {
 		capture_type = CFR_CAPTURE_METHOD_AUTO;
 	}
 
-	if (FIELD_GET(CFIR_DMA_HDR_INFO2_MURX_DATA_INC, dma_hdr.info2)) {
-		if (freeze_tlv_ver == MACRX_FREEZE_TLV_VERSION_3)
-			freeze_tlv_len = sizeof(struct macrx_freeze_capture_channel_v3);
+	if (dma_hdr.mu_rx_data_incl) {
+		if (dma_hdr.freeze_tlv_version == MACRX_FREEZE_TLV_VERSION_5)
+			freeze_tlv_len =
+				sizeof(struct macrx_freeze_capture_channel_v5);
+		else if (dma_hdr.freeze_tlv_version == MACRX_FREEZE_TLV_VERSION_3)
+			freeze_tlv_len =
+				sizeof(struct macrx_freeze_capture_channel_v3);
 		else
 			freeze_tlv_len = sizeof(struct macrx_freeze_capture_channel);
 
@@ -274,7 +275,7 @@ static int ath12k_cfr_enh_process_data(s
 				  freeze_tlv_len;
 	}
 
-	length = FIELD_GET(CFIR_DMA_HDR_INFO0_LEN, dma_hdr.hdr.info0) * 4;
+	length = dma_hdr.length * 4;
 	length += dma_hdr.total_bytes;
 
 	spin_lock_bh(&cfr->lut_lock);
@@ -299,28 +300,28 @@ static int ath12k_cfr_enh_process_data(s
 	lut->buff = param->buff;
 	lut->data = data;
 	lut->data_len = length;
-	lut->dbr_ppdu_id = dma_hdr.hdr.phy_ppdu_id;
+	lut->dbr_ppdu_id = dma_hdr.phy_ppdu_id;
 	lut->dbr_tstamp = jiffies;
-	lut->header_length = FIELD_GET(CFIR_DMA_HDR_INFO0_LEN, dma_hdr.hdr.info0);
+	lut->header_length = dma_hdr.length;
 	lut->payload_length = dma_hdr.total_bytes;
-	memcpy(&lut->dma_hdr.enh_hdr, &dma_hdr, sizeof(struct ath12k_cfir_enh_dma_hdr));
+	memcpy(&lut->dma_hdr.enh_hdr, &dma_hdr,
+	       sizeof(struct ath12k_cfir_enh_dma_hdr));
 
 	header = &lut->header;
 	meta = &header->u.meta_enh;
-	meta->channel_bw = FIELD_GET(CFIR_DMA_HDR_INFO1_UPLOAD_PKT_BW,
-				     dma_hdr.hdr.info1);
+	meta->channel_bw = dma_hdr.upload_pkt_bw;
 	meta->num_rx_chain =
-		NUM_CHAINS_FW_TO_HOST(FIELD_GET(CFIR_DMA_HDR_INFO1_NUM_CHAINS,
-						dma_hdr.hdr.info1));
+		NUM_CHAINS_FW_TO_HOST(dma_hdr.num_chains);
 	meta->length = length;
 
 	if (capture_type != CFR_CAPTURE_METHOD_ACK_RESP_TO_TM_FTM) {
 		meta->capture_type = capture_type;
-		meta->sts_count = FIELD_GET(CFIR_DMA_HDR_INFO1_NSS, dma_hdr.hdr.info1) + 1;
-		if (FIELD_GET(CFIR_DMA_HDR_INFO2_MURX_DATA_INC, dma_hdr.info2)) {
+		meta->sts_count = dma_hdr.nss + 1;
+		if (!dma_hdr.mu_rx_data_incl) {
 			peer_macaddr = meta->peer_addr.su_peer_addr;
-			if (freeze_tlv)
-				extract_peer_mac_from_freeze_tlv(freeze_tlv, peer_macaddr);
+			if (dma_hdr.freeze_data_incl)
+				extract_peer_mac_from_freeze_tlv(freeze_tlv,
+								 peer_macaddr);
 		}
 	}
 
